test/unit/format/sam_file_format_test_template.hpp

```diff
@@ -0,0 +1,724 @@
+    EXPECT_EQ(header.program_infos[0].version, "1.0.0");
+    EXPECT_EQ(header.program_infos[0].description, "trim reads with low qual");
+    EXPECT_EQ(header.program_infos[0].previous, "");
+    EXPECT_EQ(header.program_infos[0].command_line_call, "qc -f file1");
+    EXPECT_EQ(header.program_infos[1].id, "novoalign");
+    EXPECT_EQ(header.program_infos[1].name, "novoalign");
+    EXPECT_EQ(header.program_infos[1].version, "V3.02.07");
+    EXPECT_EQ(header.program_infos[1].description, "");
+    EXPECT_EQ(header.program_infos[1].previous, "qc");
+    EXPECT_EQ(header.program_infos[1].command_line_call, "novoalign -d /path/hs37d5.ndx -f /path/file.fastq.gz");
+
+    std::string id1{"ref"};
+    std::string id2{"ref2"};
+
+    EXPECT_EQ(header.ref_id_info[(header.ref_dict[id1])], (std::tuple<uint32_t, std::string>{249250621u, ""}));
+    EXPECT_EQ(header.ref_id_info[(header.ref_dict[id2])], (std::tuple<uint32_t, std::string>{243199373u, "AS:hs37d5"}));
+
+    EXPECT_EQ(header.read_groups[0],
+              (std::pair<std::string, std::string>{"U0a_A2_L1", "PL:illumina\tPU:1\tLB:1\tSM:NA12878"}));
+    EXPECT_EQ(header.read_groups[1],
+              (std::pair<std::string, std::string>{"U0a_A2_L2", "PL:illumina\tSM:NA12878\tPU:1\tLB:1"}));
+
+    EXPECT_EQ(header.comments[0], "Tralalalalalala this is a comment");
+}
+*/
+
+TYPED_TEST_P(sam_file_read, read_in_all_data)
+{
+    // prepare tag dictionary
+    this->tag_dicts[0]["NM"_tag] = -7;
+    this->tag_dicts[0]["AS"_tag] = 2;
+    this->tag_dicts[0]["CC"_tag] = 300;
+    this->tag_dicts[0]["cc"_tag] = -300;
+    this->tag_dicts[0]["aa"_tag] = 'c';
+    this->tag_dicts[0]["ff"_tag] = 3.1f;
+    this->tag_dicts[0]["zz"_tag] = "str";
+    this->tag_dicts[1]["bc"_tag] = std::vector<int8_t>{-3};
+    this->tag_dicts[1]["bC"_tag] = std::vector<uint8_t>{3u, 200u};
+    this->tag_dicts[1]["bs"_tag] = std::vector<int16_t>{-3, 200, -300};
+    this->tag_dicts[1]["bS"_tag] = std::vector<uint16_t>{300u, 40u, 500u};
+    this->tag_dicts[1]["bi"_tag] = std::vector<int32_t>{-3, 200, -66000};
+    this->tag_dicts[1]["bI"_tag] = std::vector<uint32_t>{294967296u};
+    this->tag_dicts[1]["bf"_tag] = std::vector<float>{3.5f, 0.1f, 43.8f};
+    this->tag_dicts[1]["bH"_tag] = std::vector<std::byte>{std::byte{0x1A}, std::byte{0xE3}, std::byte{0x01}};
+
+    // read data
+    typename TestFixture::stream_type istream{this->verbose_reads_input};
+    using record_t =
+      bio::detail::record_from_typelist<std::remove_cvref_t<decltype(bio::map_io::default_field_ids)>,
+                                        std::remove_cvref_t<decltype(bio::map_io::field_types_sam<>)>>;
+    bio::map_io::reader_options default_options{};
+    bio::format_input_handler<bio::sam> input_handler{istream, default_options};
+    record_t rec;
+
+    for (unsigned i = 0; i < 3; ++i)
+    {
+        input_handler.parse_next_record_into(rec);
+
+        EXPECT_EQ(rec.id(), this->ids[i]);
```

Comment:
@h-2 So my first test almost passes except that it only reads the first two records correctly and then the second read again so my test fails. I dhave no clue why.